source/initial_temperature/patch_on_S40RTS.cc

```diff
@@ -0,0 +1,204 @@
+                                            s40rts.reference_temperature;
+
+      // get the vs to density conversion
+      const double depth = this->get_geometry_model().depth(position);
+
+      double vs_to_density = 0.0;
+      if (s40rts.vs_to_density_method == s40rts.file)
+        vs_to_density = s40rts.profile.get_data_component(Point<1>(depth), s40rts.vs_to_density_index);
+      else if (s40rts.vs_to_density_method == s40rts.constant)
+        vs_to_density = s40rts.vs_to_density_constant;
+      else
+        // we shouldn't get here but instead should already have been
+        // kicked out by the assertion in the parse_parameters()
+        // function
+        Assert (false, ExcNotImplemented());
+
+      // scale the perturbation in seismic velocity into a density perturbation
+      // vs_to_density is an input parameter
+      const double density_perturbation = vs_to_density * vs_perturbation;
+
+      double temperature_perturbation;
+      if (depth > s40rts.no_perturbation_depth)
+        // scale the density perturbation into a temperature perturbation
+        temperature_perturbation =  -1./s40rts.thermal_alpha * density_perturbation;
+      else
+        // set heterogeneity to zero down to a specified depth
+        temperature_perturbation = 0.0;
+
+      return background_temperature + temperature_perturbation;
+    }
+
+
+
+
+    template <int dim>
+    void
+    PatchOnS40RTS<dim>::declare_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection ("Initial temperature model");
+      {
+        Utilities::AsciiDataBase<dim>::declare_parameters(prm,
+                                                          "$ASPECT_SOURCE_DIR/data/initial-temperature/patch-on-S40RTS/test/",
+                                                          "upper_shell_3d.txt");
+        prm.declare_entry ("Maximum grid depth", "0.0",
+                           Patterns::Double (0),
+                           "The maximum depth of the Vs ascii grid. The model will read in  "
+                           "Vs from S40RTS below this depth.");
+        prm.declare_entry ("Smoothing length scale", "200000.0",
+                           Patterns::Double (0),
+                           "The depth range (above maximum grid depth) over which to smooth. "
+                           "Smoothing is done by averaging the ascii grid and S40RTS at each point.");
+
+      }
+      prm.leave_subsection();
+    }
+
+
+
+    template <int dim>
+    void
+    PatchOnS40RTS<dim>::parse_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection("Initial temperature model");
+      {
+        Utilities::AsciiDataBase<dim>::parse_parameters(prm);
+
+        max_grid_depth   = prm.get_double ("Maximum grid depth");
+        smoothing_length_scale   = prm.get_double ("Smoothing length scale");
+
+        AssertThrow (max_grid_depth != 0,
```

Comment:
What is the reason for max_grid_depth not being allowed to be zero? Is it allowed to be negative?
Instead of testing for not being zero, maybe test that the max_grid_depth is bigger than the smoothing length?

Reply:
Here I was trying to make it so the user has to declare a maximum grid depth - I thought it didn't make that much sense to have a default when I have no idea what size grids people will use. Perhaps there is a better way to do this?